jupyterhub/handlers/base.py

```diff
@@ -425,14 +427,35 @@ async def get_current_user(self):
                 if user and isinstance(user, User):
                     user = await self.refresh_auth(user)
                 self._jupyterhub_user = user
+                self._parse_scopes()
             except Exception:
                 # don't let errors here raise more than once
                 self._jupyterhub_user = None
                 self.log.exception("Error getting current user")
-        if self._jupyterhub_user is not None or self.get_current_user_oauth_token():
-            self.scopes = roles.get_subscopes(*self._jupyterhub_user.roles)
+        self._parse_scopes()
+        self._cache_db()
         return self._jupyterhub_user
 
+    def _cache_db(self):
+        self.db_cache = {}
+        for kind in {'users', 'groups', 'services'}:
+            Resource = orm.get_class(kind)
+            resources_names = {
+                resource.name for resource in self.db.query(Resource).all()
```

Comment:
This preemptively fetches *all* resources from the database into memory on every request. I think that will be too expensive.

I think the idea is, within a given request i.e. in `Handler.get()`, not in the base handler preparation), fetch the resources that will be relevant for that request. That should be, e.g. Users (possibly filtered) and fully resolved scopes for the token. Then, when it comes time to do filtering, we can do that with the in-memory objects we have already retrieved. These should all already be available in the resolved_scopes and orm.User object, for instance.


I think the gist should be something like:

```python
def check_scope(resource, need_scopes, have_scopes):
    if it's a user and I have !user=user.name:
        return True
    if it's a group and I have !group=group.name:
       return True
    if it's a user and I have any(!group=group.name for group in user.groups):
       return True
```

None of these construct new queries, and we shouldn't need to fetch users or groups as part of constructing the filter - the filter should only be fetching roles -> scopes, and *applying* the filter, should work on local attributes only. No additional lookups, needed.